src/nvim/charset.c

```diff
@@ -1621,131 +1621,94 @@ bool vim_isblankline(char_u *lbuf)
 void vim_str2nr(const char_u *const start, int *const prep, int *const len,
                 const int what, varnumber_T *const nptr,
                 uvarnumber_T *const unptr, const int maxlen)
+  FUNC_ATTR_NONNULL_ARG(1)
 {
-  const char_u *ptr = start;
+  const char *ptr = (const char *)start;
+#define STRING_ENDED(ptr) \
+    (!(maxlen == 0 || (int)((ptr) - (const char *)start) < maxlen))
   int pre = 0;  // default is decimal
   bool negative = false;
-  uvarnumber_T un = 0;
 
   if (ptr[0] == '-') {
     negative = true;
     ptr++;
   }
 
   // Recognize hex, octal and bin.
-  if ((ptr[0] == '0') && (ptr[1] != '8') && (ptr[1] != '9')
-      && (maxlen == 0 || maxlen > 1)) {
+  if ((what & (STR2NR_HEX|STR2NR_OCT|STR2NR_BIN))
+      && !STRING_ENDED(ptr + 1)
+      && ptr[0] == '0' && ptr[1] != '8' && ptr[1] != '9') {
     pre = ptr[1];
 
     if ((what & STR2NR_HEX)
-        && ((pre == 'X') || (pre == 'x'))
-        && ascii_isxdigit(ptr[2])
-        && (maxlen == 0 || maxlen > 2)) {
+        && !STRING_ENDED(ptr + 2)
+        && (pre == 'X' || pre == 'x')
+        && ascii_isxdigit(ptr[2])) {
       // hexadecimal
       ptr += 2;
     } else if ((what & STR2NR_BIN)
-               && ((pre == 'B') || (pre == 'b'))
-               && ascii_isbdigit(ptr[2])
-               && (maxlen == 0 || maxlen > 2)) {
+               && !STRING_ENDED(ptr + 2)
+               && (pre == 'B' || pre == 'b')
+               && ascii_isbdigit(ptr[2])) {
       // binary
       ptr += 2;
     } else {
       // decimal or octal, default is decimal
       pre = 0;
 
-      if (what & STR2NR_OCT) {
-        // Don't interpret "0", "08" or "0129" as octal.
-        for (int n = 1; ascii_isdigit(ptr[n]); ++n) {
-          if (ptr[n] > '7') {
-            // can't be octal
+      if (what & STR2NR_OCT
```

Comment:
If `STR2NR_FORCE` is also set, we later on always parse the string in base `8`. So checking for invalid digits would be unnecessary here.